Support FETtecOneWire ESC protocol (with telemetry)

amilcarlucas opened the pull request:
This is a 2Mbaud UART based bi-directional protocol.
The initial code version was provided directly by FETtec

I need some help integrating this because there is no UART based ESCs that I know of.
The protocol supports up-to 25 ESC using a single UART and a single wire.

FETtec will provide hardware to anyone willing to help on this one.

peterbarker commented:
@amilcarlucas I'd like to get one of those for testing if I could.

amilcarlucas commented:
Sure, can you send me your private address per PM or e-mail ?

amilcarlucas commented:
Anything I can do to help here?

Linjieqiang commented:
Any news?

khancyr commented:
I will come, I need a little time to sort it out

Linjieqiang commented:
@khancyr Nice.

Linjieqiang commented:
@amilcarlucas I found that the serial port baud rate of many MCUs is up to less than 2M. So can the baud rate setting of this protocol be reduced?

amilcarlucas commented:
@teezeefpv this is the PR, please add any changes here.
You also have permissions to push changes to my repository, and therefore update this PR

amilcarlucas commented:
@andyp1per I guess in the end this will have a lot of similarities with the bidirectional dhot stuff you did.

teezeefpv commented:
Hi@all. Thank you for your support. We wrote some documentation. It is not perfect yet, but it might help. 
[FETtec OW manual.pdf](https://github.com/ArduPilot/ardupilot/files/5924773/FETtec.OW.manual.pdf)

amilcarlucas commented:
@teezeefpv do you have a newer version of the code? if you do [please update it](https://ardupilot.org/dev/docs/editing-prs.html)

khancyr commented:
@amilcarlucas I got the latest lib, I just didn't had time to put it into ardupilot tree ...

teezeefpv commented:
> @teezeefpv do you have a newer version of the code? if you do [please update it](https://ardupilot.org/dev/docs/editing-prs.html)

I just removed unused includes. There is no newer version atm

teezeefpv commented:
> @amilcarlucas I found that the serial port baud rate of many MCUs is up to less than 2M. So can the baud rate setting of this protocol be reduced?

It would be possible to change the baudrate, but we would need to integrate a robust, dynamic baudrate detection into our bootloader, that hasn't enough space left. We keep that in mind, but this will take some time to find a good solution. I would like to try 2Mbaud with Ardupilot. Is it correct, that it would be an additional case in this file or is there more?
https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_SerialManager/AP_SerialManager.cpp#L613

We did some research and testing the last days, trying to find out why and if there would be reasons to reduce the baudrate. There seems to be a lot of historical statements, where (presumably) the edge steepness wasn't as good as it is today with modern transistors. This statemens say, that the cable length can not be longer than a few centimetres (~15cm).
We tested 2Mbaud with four meters of unshielded wire without problems. Even when we wrap the cable around the electronic no errors occured, although there was noise visible on the oscilloscope. 

In my understanding this steps are needed to get it up and running:
1. Rise the Baudrate to 2000000
2. Find the container/functions for serial communication -> Init, Deinit, AvailableBytes, Read and Write Bytes
3. Find the container/functions that are needed for ESC values and telemetry @andyp1per I would love to have a chat about that
4. integrate the onewire lib by changing the #defines in the h file. 

Am I missing something?

khancyr commented:
I have made a first compiling version.
We support 2M as baudrate already, it is just not in the list. As that is not variable I have hard set it.
I have made the related changes for uart usage.

I need to implement the mavlink esc reporting and the library usage but nothing hard.

Then we will need some testing

amilcarlucas commented:
Just out of curiosity, how much extra memory does this use on a CubeOrange ?

amilcarlucas commented on libraries/AP_SerialManager/AP_SerialManager.h:
```diff
@@ -150,6 +150,7 @@ class AP_SerialManager {
         SerialProtocol_AirSpeed = 34,
         SerialProtocol_ADSB = 35,
         SerialProtocol_AHRS = 36,
+        SerialProtocol_FETtechOneWire = 37,
```
This also needs to be added to the parameter values list, right?

khancyr replied:
yep.

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+    FETtecOneWire_ResponseLength[OW_SET_ID] = 1;
+
+/*
+    FETtecOneWire_ResponseLength[OW_GET_LINEAR_THRUST] = 1;
```
@teezeefpv why are these commented out?

teezeefpv replied:
That wasn't a great benefit yet

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+                mavlink_msg_esc_telemetry_1_to_4_send((mavlink_channel_t)mav_chan, temperature, voltage, current, totalcurrent, rpm, count);
+            } else {
+                mavlink_msg_esc_telemetry_5_to_8_send((mavlink_channel_t)mav_chan, temperature, voltage, current, totalcurrent, rpm, count);
+            }
```
Don't we have a 9_to12 message already?

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+        FETtecOneWire_FoundESCs = 0;
+        FETtecOneWire_ScanActive = 0;
+        FETtecOneWire_SetupActive = 0;
+        FETtecOneWire_minID = 25;
```
How about using a static const int, instead of the hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+        FETtecOneWire_maxID = 0;
+        FETtecOneWire_IDcount = 0;
+        FETtecOneWire_FastThrottleByteCount = 0;
+        for (uint8_t i = 0; i < 25; i++) {
```
Again the hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+{
+    /*
+    a frame lookes like:
+    byte 1 = fremae header (master is always 0x01)
```
s/fremae/frame/

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+}
+
+/*
+    scans for ESC's in bus. should be called intill FETtecOneWire_ScanActive >= 25
```
s/intill/untill

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+}
+
+/*
+    starts all ESC's in bus and prepares them for receiving teh fast throttle command should be called untill FETtecOneWire_SetupActive >= 25
```
s/teh/the/

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+        wakeFromBL = 1;
+        return FETtecOneWire_SetupActive + 1;
+    }
+    while (FETtecOneWire_activeESC_IDs[FETtecOneWire_SetupActive] == 0 && FETtecOneWire_SetupActive < 25) {
```
Again hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+        FETtecOneWire_SetupActive++;
+    }
+
+    if (FETtecOneWire_SetupActive == 25 && wakeFromBL == 0) {
```
hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+
+    if (FETtecOneWire_SetupActive == 25 && wakeFromBL == 0) {
+        return FETtecOneWire_SetupActive;
+    } else if (FETtecOneWire_SetupActive == 25 && wakeFromBL) {
```
hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+        State = 0;
+        TimeOut = 0;
+
+        FETtecOneWire_minID = 25;
```
Hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,850 @@
+        TimeOut = 0;
+
+        FETtecOneWire_minID = 25;
+        FETtecOneWire_maxID = 0;
```
Hardcoded 25

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.h:
```diff
@@ -0,0 +1,274 @@
+    uint8_t FETtecOneWire_PullCommand(uint8_t ESC_id, uint8_t *command, uint8_t *response, uint8_t returnFullFrame);
+
+/*
+    scans for ESC's in bus. should be called intill FETtecOneWire_ScanActive >= 25
```
s/intill/untill

amilcarlucas reviewed (commented):
small stuff from a fast scan. I will submit a PR later if no one beats me to it.

khancyr commented:
@amilcarlucas I mostly didn't touch the fettech example code for now, I think we should only wrap it for now to get everything work. And after we could refactor it in something more AP and c++ friendly.

amilcarlucas commented:
@teezeefpv can you test it now?

amilcarlucas commented:
SITL is compiling fine, but other targets are not :(

amilcarlucas commented:
To test this (using for instance serial port 2, and a quadcopter) set these parametrs:
SERIAL2_PROTOCOL = 37
SERVO_FTW_MASK = 15

Then connect the ESCs to serial port 2 and reboot the system.

amilcarlucas commented:
@teezeefpv please improve:

-  ArduPilot/ardupilot_wiki#3333
-  ArduPilot/ardupilot_wiki#3334
-  ArduPilot/ardupilot_wiki#3335

amilcarlucas commented:
Rebased on master because it contains a half-duplex fix, fixed documentation a bit more, removed unused functions, move static variables into member variables, added a static assert ... and force pushed it.

amilcarlucas commented on libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:
```diff
@@ -0,0 +1,807 @@
+
+AP_FETtecOneWire *AP_FETtecOneWire::_singleton;
+
+static uint8_t FETtecOneWire_ResponseLength[54];
```
@khancyr Is there a reason for not making these two variables be member variables?

amilcarlucas commented:
I did some changes here, and in ArduPilot/ardupilot_wiki#3335 as well

amilcarlucas reviewed (commented):
Thanks, but you should delete the questions. Those were just there as a guideline.
The idea is to provide the information in a "Fließtext" Format, not a question-Answer format.

amilcarlucas commented:
Version 7be6e6f was tested today in hardware, the motors respond, and the telemetry looks ok.
A initial flight test with a hexacopter will be done soon.

amilcarlucas commented:
@teezeefpv any updates on this? How is the documentation comming?

teezeefpv commented:
As OneWire is associated with the SERVO*_FUNCTIONs only eight motors can be used right? Parameter 33 is first motor and 40 is motor 8. Would we need to add additional functions to get 12 motors working?

amilcarlucas commented:
Yes:

82 | Motor9
83 | Motor10
84 | Motor11
85 | Motor12

amilcarlucas commented:
6550364c was flight tested successfully by FETtec on an Hexacopter

andyp1per commented:
Thanks @amilcarlucas - the logging change, that was fixed in a rush by me - so does this fix my fix or something else?

amilcarlucas commented:
I am preparing a PR to your branch as we speak. We can discuss it there

amilcarlucas commented:
@andyp1per I created https://github.com/andyp1per/ardupilot/pull/15 we can discuss it there

amilcarlucas commented:
Testing uncovered some issues with Pixhawk1. CubeBlack is working fine. So we are still investigating this.

amilcarlucas commented:
Superseded by #16850 and #16874
